Doc/library/asyncio-eventloop.rst

```diff
@@ -966,6 +966,11 @@ Unix signals
 
    Set *callback* as the handler for the *signum* signal.
 
+   The callback will be invoked in the thread that runs *loop*, along with
```

Comment:
"in the thread that runs *loop*"  it's worth clarifying that using this function is only allowed from the main thread.  The loop that sets up signal handler must later only operate in the main loop too.

Reply:
Is the second part really true? Python guarantees that the signal handler will be invoked by the main thread, but `_UnixSelectorEventLoop._process_self_data` should be able to pick up the bytes written into the pipe regardless of which thread the event loop runs in?

Reply:
> Is the second part really true?

No.  It will probably work alright if you have just one loop.  But what if you have two?  You set up signal handlers for both of them while in the main thread, if those signals overlap then the first loop won't receive them.

Frankly, the current signals support in asyncio is far from ideal.  That's why I'd like to focus the docs on what's actually proven to work correctly.

Reply:
Fair enough. If you want to document that `add_signal_handler` won't work if different event loops subscribe to the same signal and run at the same time, I can add a sentence to that effect. (Though it does seem a bit esoteric for this level of documentation.)

On the other hand, a single event loop running in a dedicated thread is a legitimate use of asyncio (and maybe even _the_ way to introduce asyncio to legacy programs), so I'd prefer not to document it as unsupported unless there's a reason why it doesn't or wouldn't work.

Reply:
`add_signal_handler()` calls `signal.set_wakeup_fd()` which in turn is only allowed to be executed in the main thread.

@1st1 is 100% right. I think the doc should mention the main thread only.

Reply:
@asvetlov Agreed that `add_signal_handler()` may only be called in the main thread, no argument there. But once you set up the handler (in the main thread), there is no reason why your event loop wouldn't be allowed to _run_ in a background thread. As already mentioned, `_process_self_data` should be able to pick up the bytes written into the pipe regardless of the thread the event loop runs in.

The proposed wording only states where the callback will be executed. It could also mention that `add_signal_handler` itself must be called from the main thread. Does that sound reasonable?

Reply:
Are you talking about creating a loop from the main thread but executing it from a different thread?
It looks like antipattern for me.
Honestly, I don't know if asyncio even supports such trick; but even it does -- the behavior can be changed in future releases.

Reply:
> On the other hand, a single event loop running in a dedicated thread is a legitimate use of asyncio (and maybe even the way to introduce asyncio to legacy programs)

Do you use asyncio that way yourself?  Or does some popular framework use it that way?

Reply:
The actual use case I had in mind is just running the event loop in a dedicated thread.

Creating the loop in the main thread and running elsewhere would only be necessary to work around the fact that `add_signal_handler` must be invoked in the main thread. If you are saying that `new_event_loop` and `run_until_complete` and friends must be run in the same thread, then I guess `add_signal_handler` is truly not supported in event loops that don't run in the main thread.

But that limitation is not obvious to me and I couldn't find it in the documentation. What is the reason that it wouldn't work to create an event loop in one thread (without ever running it) and run it in another?

Reply:
@1st1 Not yet (other than in tests and toy examples), but I plan to introduce asyncio into a large application where the main thread runs the GUI main loop. Asyncio would run in a separate thread, and the GUI would communicate with it using thread-safe entry points such as `asyncio.run_coroutine_threadsafe` and `loop.call_soon_threadsafe`.

Just to clarify, I have no need for `add_signal_handler` in that scenario, I merely wanted to document how it differs from `signal.signal` after the question on StackOverflow.

Reply:
> But that limitation is not obvious to me and I couldn't find it in the documentation. What is the reason that it wouldn't work to create an event loop in one thread (without ever running it) and run it in another?

Suppose you have two loops, L1 and L2.

```python
# in the main thread:
L1.add_signal_handler(SIGINT, l1_handler)
L2.add_signal_handler(SIGINT, l2_handler)
```

OR

```python
# in the main thread:
L1.add_signal_handler(SIGINT, l1_handler)
# later in the main thread in some third-party blocking IO dep:
signal.signal(SIGINT, ...)
```

IIRC basically in both of the above scenarios L1 won't receive `SIGINT` anymore.

The only "safe" scenario is therefore to use asyncio in the main thread and nothing else.  No blocking IO libraries in other threads, no other asyncio threads, no alternative async/await libraries in parallel threads.  That said, *it is* possible to do what you want too!  But it's not safe and can lead to some very hard to debug bugs.  That's why @asvetlov and I are so careful about documenting this stuff -- better to say what's right so that people follow, than to be vague and let people down.

Reply:
@1st1 Thanks for the explanation. I was referring to the case where a *single* event loop calls `add_signal_handler` in the main thread and is then run in another thread. But it's a stretch, so I'll just remove the mention of threads from the PR to spare us from further discussion in which we agree about all the essentials.

To clarify my question about "that limitation": I understood @asvetlov to *generally* claim that it's not allowed to create an event loop in one thread and run it in another, even if it doesn't touch signals.

Reply:
> @1st1 Thanks for the explanation. I was referring to the case where a single event loop calls add_signal_handler in the main thread and is then run in another thread.

Yes, I got that.

> But it's a stretch, so I'll just remove the mention of threads from the PR to spare us from further discussion in which we agree about all the essentials.

Exactly, it's a narrow case that you can manage to get working if you're careful.  Not worthy of bending documentation to explain it.